File: android/src/main/kotlin/io/rodcone/flutter_meta_wearables_dat/DeviceStateStreamHandler.kt
package io.rodcone.flutter_meta_wearables_dat

import android.util.Log
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.DeviceSelector
import com.meta.wearable.dat.core.types.DeviceState
import com.meta.wearable.dat.core.types.ThermalLevel
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * Stream handler for per-device state updates (currently: thermal level).
 *
 * 0.7.0 added `Wearables.getDeviceState(deviceIdentifier): StateFlow<DeviceState>`
 * keyed by `DeviceIdentifier`. The plugin's Dart-facing API exposes a single
 * `deviceStateStream()` that tracks the *active* device, so this handler
 * wraps the per-device flow in an outer subscription to
 * `deviceSelector.activeDeviceFlow()` and switches its inner subscription
 * whenever the active device changes.
 *
 * Mirrors the iOS `DeviceStateStreamHandler.swift` design exactly so the
 * Dart-side `Stream<DeviceState>` API behaves identically on both platforms.
 */
internal class DeviceStateStreamHandler(
        private val deviceSelectorProvider: () -> DeviceSelector,
        private val isInitialized: () -> Boolean,
) : EventChannel.StreamHandler {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var outerJob: Job? = null
    private var innerJob: Job? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (events == null) return
        eventSink = events
        if (!isInitialized()) {
            // SDK not yet initialized — bail. The plugin can call
            // [restartMonitoring] once Bluetooth permissions are granted and
            // `Wearables.initialize` has run.
            Log.d(TAG, "DeviceStateStream onListen — SDK not initialized yet")
            return
        }
        startCollecting(events)
    }

    override fun onCancel(arguments: Any?) {
        outerJob?.cancel()
        outerJob = null
        innerJob?.cancel()
        innerJob = null
        eventSink = null
    }

    /**
     * Restart the active-device subscription. Called after BT permissions are
     * granted so we don't miss the first device emission.
     */
    fun restartMonitoring() {
        val sink = eventSink ?: return
        if (outerJob?.isActive == true) return
        if (!isInitialized()) return
        startCollecting(sink)
    }

    private fun startCollecting(events: EventChannel.EventSink) {
        outerJob?.cancel()
        innerJob?.cancel()
        innerJob = null
        val selector = deviceSelectorProvider()
        var currentDeviceId: com.meta.wearable.dat.core.types.DeviceIdentifier? = null

        // Seed: if a device is already active, attach the inner subscription
        // immediately so Dart subscribers see the first thermal reading
        // without waiting for an `activeDeviceFlow` tick.
        selector.activeDevice()?.let { deviceId ->
            currentDeviceId = deviceId
            innerJob = subscribe(deviceId, events)
        }

        outerJob =
                scope.launch {
                    selector.activeDeviceFlow().collect { deviceId ->
                        // `activeDeviceFlow()` replays the current value to
                        // new collectors. The SDK's
                        // `Wearables.getDeviceState(...)` doesn't tolerate
                        // rapid cancel+resubscribe for the same device, so
                        // only tear down + restart when the device actually
                        // changes (mirrors the iOS handler's behaviour).
                        if (deviceId == currentDeviceId) return@collect
                        innerJob?.cancel()
                        innerJob = null
                        currentDeviceId = deviceId
                        if (deviceId != null) {
                            innerJob = subscribe(deviceId, events)
                        }
                    }
                }
    }

    private fun subscribe(
            deviceId: com.meta.wearable.dat.core.types.DeviceIdentifier,
            events: EventChannel.EventSink,
    ): Job {
        return scope.launch {
            Wearables.getDeviceState(deviceId).collect { state ->
                events.success(mapOf("thermalLevel" to thermalLevelToInt(state.thermalLevel)))
            }
        }
    }

    fun dispose() {
        outerJob?.cancel()
        outerJob = null
        innerJob?.cancel()
        innerJob = null
        eventSink = null
        scope.cancel()
    }

    companion object {
        private const val TAG = "MetaWearablesDat"

        /**
         * Mirrors the int values the Dart `ThermalLevel` enum uses
         * (also the same as the iOS handler emits). Keep these aligned.
         */
        private fun thermalLevelToInt(level: ThermalLevel): Int {
            return when (level) {
                ThermalLevel.UNKNOWN -> 0
                ThermalLevel.NONE -> 1
                ThermalLevel.LIGHT -> 2
                ThermalLevel.MODERATE -> 3
                ThermalLevel.SEVERE -> 4
                ThermalLevel.CRITICAL -> 5
                ThermalLevel.EMERGENCY -> 6
                ThermalLevel.SHUTDOWN -> 7
            }
        }
    }
}

File: ios/flutter_meta_wearables_dat/Sources/flutter_meta_wearables_dat/DeviceStateStreamHandler.swift
import Flutter
import MWDATCore

/// Observes the selected device's thermal state through DAT 1.0 Device listeners.
class DeviceStateStreamHandler: NSObject, FlutterStreamHandler {
  private let deviceSelectorProvider: @MainActor () -> AutoDeviceSelector
  private var outerTask: Task<Void, Never>?
  private var listenerToken: AnyListenerToken?
  private var eventSink: FlutterEventSink?
  private var isMonitoring = false
  private var monitoringGeneration = 0
  private var deviceGeneration = 0

  init(deviceSelectorProvider: @escaping @MainActor () -> AutoDeviceSelector) {
    self.deviceSelectorProvider = deviceSelectorProvider
    super.init()
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    startMonitoring()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    monitoringGeneration &+= 1
    outerTask?.cancel()
    outerTask = nil
    cancelDeviceListener()
    eventSink = nil
    isMonitoring = false
    return nil
  }

  func restartMonitoring(force: Bool = false) {
    guard eventSink != nil else { return }
    if !force, isMonitoring { return }
    startMonitoring()
  }

  private func cancelDeviceListener() {
    // Cancellation is async: invalidate queued callbacks before scheduling it.
    deviceGeneration &+= 1
    if let token = listenerToken {
      listenerToken = nil
      Task { await token.cancel() }
    }
  }

  private func startMonitoring() {
    outerTask?.cancel()
    cancelDeviceListener()
    guard eventSink != nil else { return }
    monitoringGeneration &+= 1
    let generation = monitoringGeneration
    isMonitoring = true
    outerTask = Task { @MainActor in
      defer {
        if self.monitoringGeneration == generation {
          self.isMonitoring = false
          self.cancelDeviceListener()
        }
      }
      guard !Task.isCancelled, self.monitoringGeneration == generation else { return }
      let selector = self.deviceSelectorProvider()
      var currentDeviceId = selector.activeDevice
      if let deviceId = currentDeviceId { self.subscribe(toDevice: deviceId) }
      for await deviceId in selector.activeDeviceStream() {
        guard !Task.isCancelled, self.monitoringGeneration == generation else { return }
        if deviceId == currentDeviceId { continue }
        self.cancelDeviceListener()
        currentDeviceId = deviceId
        if let deviceId = deviceId { self.subscribe(toDevice: deviceId) }
      }
    }
  }

  @MainActor
  private func subscribe(toDevice deviceId: DeviceIdentifier) {
    guard let device = Wearables.shared.deviceForIdentifier(deviceId) else { return }
    let generation = deviceGeneration
    // The SDK delivers a snapshot immediately, then on each device-state change.
    listenerToken = device.addDeviceStateListener { [weak self] state in
      Task { @MainActor in
        guard let self, self.deviceGeneration == generation else { return }
        self.eventSink?(Self.stateToMap(state))
      }
    }
  }

  private static func stateToMap(_ state: DeviceState) -> [String: Any] {
    return ["thermalLevel": thermalLevelToInt(state.thermalLevel)]
  }

  /// Mirrors the int values exposed by the Dart `ThermalLevel` enum.
  private static func thermalLevelToInt(_ level: ThermalLevel) -> Int {
    switch level {
    case .unknown:    return 0
    case .none:       return 1
    case .light:      return 2
    case .moderate:   return 3
    case .severe:     return 4
    case .critical:   return 5
    case .emergency:  return 6
    case .shutdown:   return 7
    @unknown default: return 0
    }
  }
}
